addons/hw_drivers/tools/helpers.py

```diff
@@ -197,17 +198,19 @@ def generate_password():
     """
     Generate an unique code to secure raspberry pi
     """
-    alphabet = 'abcdefghijkmnpqrstuvwxyz23456789'
-    password = ''.join(secrets.choice(alphabet) for i in range(12))
-    try:
-        shadow_password = crypt.crypt(password, crypt.mksalt())
-        subprocess.run(('sudo', 'usermod', '-p', shadow_password, 'pi'), check=True)
-        with writable():
-            subprocess.run(('sudo', 'cp', '/etc/shadow', '/root_bypass_ramdisks/etc/shadow'), check=True)
-        return password
-    except subprocess.CalledProcessError as e:
-        _logger.error("Failed to generate password: %s", e.output)
-        return 'Error: Check IoT log'
+    if main.ssh_service:
+        alphabet = 'abcdefghijkmnpqrstuvwxyz23456789'
+        password = ''.join(secrets.choice(alphabet) for i in range(12))
+        try:
+            shadow_password = crypt.crypt(password, crypt.mksalt())
+            subprocess.run(('sudo', 'usermod', '-p', shadow_password, 'pi'), check=True)
+            with writable():
+                subprocess.run(('sudo', 'cp', '/etc/shadow', '/root_bypass_ramdisks/etc/shadow'), check=True)
+            return password
+        except subprocess.CalledProcessError as e:
+            _logger.error("Failed to generate password: %s", e.output)
```

Comment:
any reason that you didn't used _logger.exception here ?

Reply:
This was already the case before this commit, this is not my code, Do you think I should change it ?

Reply:
@viforget yes in my opinion. The traceback and full exception error details might be useful

Reply:
@viforget it looks correct, in this case, e.output should be kept as - I think - it will not be given explicitly in the _logger.exception by default (as output looks to be a custom parameter meant only for this exception:
https://docs.python.org/3/library/subprocess.html#subprocess.CalledProcessError.output